src/CommunityToolkit.Maui/Views/Popup/PopupContainer.cs

```diff
@@ -0,0 +1,149 @@
+		: this(view as Popup<T> ?? CreatePopupFromView<Popup<T>>(view), popupOptions, taskCompletionSource)
+	{
+	}
+
+	public PopupContainer(Popup<T> popup, PopupOptions popupOptions, TaskCompletionSource<PopupResult<T>> taskCompletionSource)
+		: base(popup, popupOptions, null)
+	{
+		this.taskCompletionSource = taskCompletionSource;
+
+		Shell.SetPresentationMode(this, PresentationMode.ModalNotAnimated);
+		On<iOS>().SetModalPresentationStyle(UIModalPresentationStyle.OverFullScreen);
+	}
+
+	public Task Close(PopupResult<T> result, CancellationToken token = default)
+	{
+		token.ThrowIfCancellationRequested();
+
+		taskCompletionSource.SetResult(result);
+		return Navigation.PopModalAsync(false).WaitAsync(token);
+	}
+}
+
+partial class PopupContainer : ContentPage
+{
+	readonly Popup popup;
+	readonly PopupOptions popupOptions;
+	readonly Command tapOutsideOfPopupCommand;
+	readonly TaskCompletionSource<PopupResult>? taskCompletionSource;
+
+	public PopupContainer(View view, PopupOptions popupOptions, TaskCompletionSource<PopupResult>? taskCompletionSource)
+		: this(view as Popup ?? CreatePopupFromView<Popup>(view), popupOptions, taskCompletionSource)
+	{
+		// Only set the content if overloaded constructor hasn't set the content already; don't override content if it already exists
+		base.Content ??= new PopupContainerContent(view, popupOptions);
+	}
+
+	public PopupContainer(Popup popup, PopupOptions popupOptions, TaskCompletionSource<PopupResult>? taskCompletionSource)
+	{
+		this.popup = popup;
+		this.popupOptions = popupOptions;
+		this.taskCompletionSource = taskCompletionSource;
+
+		// Only set the content if overloaded constructor hasn't set the content already; don't override content if it already exists
+		base.Content ??= new PopupContainerContent(popup, popupOptions);
+		BackgroundColor = popupOptions.BackgroundColor;
+
+		tapOutsideOfPopupCommand = new Command(async () =>
+		{
+			popupOptions.OnTappingOutsideOfPopup?.Invoke();
+			await Close(new PopupResult(true));
+		}, () => popupOptions.CanBeDismissedByTappingOutsideOfPopup);
+
+		Content.GestureRecognizers.Add(new TapGestureRecognizer { Command = tapOutsideOfPopupCommand });
+
+		this.popupOptions.PropertyChanged += HandlePopupPropertyChanged;
+
+		this.SetBinding(BindingContextProperty, static (View x) => x.BindingContext, source: Content, mode: BindingMode.OneWay);
+
+		Shell.SetPresentationMode(this, PresentationMode.ModalNotAnimated);
+		On<iOS>().SetModalPresentationStyle(UIModalPresentationStyle.OverFullScreen);
+	}
+
+	// Prevent Content from being set by external class
+	// Casts `PopupContainer.Content` to return typeof(PopupContainerContent)
+	private protected new PopupContainerContent Content => (PopupContainerContent)base.Content;
+
+	public Task Close(PopupResult result, CancellationToken token = default)
+	{
+		token.ThrowIfCancellationRequested();
+
+		taskCompletionSource?.SetResult(result);
```

Comment:
We didn't handle double click, as I mentioned in my previous review. This will throw an exception. I think we should use `TrySetResult` instead.

Reply:
Good catch! 

```suggestion
		taskCompletionSource?.TrySetResult(result);
```